### PR TITLE
Bad UUID handling

### PR DESCRIPTION
**Pull request by InnocentBug**
# Description

## Checklist

- [x] My name is on the list of contributors (`CONTRIBUTORS.md`) in the pull request source branch.
- [x] I have updated the documentation to reflect my changes.

### CODE REVIEW

File: src/cript/api/paginator.py [hunk 1]
```diff
@@ -193,11 +193,17 @@ def fetch_page_from_api(self) -> List[dict]:
...
-        if "result" in response["data"]:
+        try:
             self.current_page_results = response["data"]["result"]
-        else:
+        except KeyError:
+            self.current_page_results = response["data"]
+        except TypeError:
             self.current_page_results = response["data"]
 
+        if response["code"] == 404 and response["error"] == "The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.":
```
Comment (nh916) [inline src/cript/api/paginator.py:203]: do you want to put the paginator refactoring in a different PR?

Comment (nh916) [reply src/cript/api/paginator.py:203]: has this been tested with the search test functions?

File: src/cript/api/api.py [hunk 2]
```diff
@@ -511,22 +510,42 @@ def save(self, project: Project) -> None:
...
-        None
+        A set of extra saved node UUIDs.
             Just sends a `POST` or `Patch` request to the API
         """
+        if known_uuid is None:
+            known_uuid = set()
 
-        project.validate()
-
+        node.validate()
```
Comment (nh916) [inline src/cript/api/api.py:519]: I see we swapped project to node. I had a similar idea where I still have the same save for the project only, but there is a helper method that takes just a node and saves that to the API, so we can reuse the code within our brute force save method

File: src/cript/api/api.py [hunk 3]
```diff
@@ -511,22 +510,42 @@ def save(self, project: Project) -> None:
...
         # saves all the local files to cloud storage right before saving the Project node
         # Ensure that all file nodes have uploaded there payload before actual save.
-        for file_node in project.find_children({"node": ["File"]}):
+        for file_node in node.find_children({"node": ["File"]}):
```
Comment (nh916) [inline src/cript/api/api.py:522]: I think if we `use a send_post_request(node)` helper method then we can leave the current code as is and the user only thinks they are saving a project still

File: src/cript/api/api.py [hunk 4]
```diff
@@ -511,22 +510,42 @@ def save(self, project: Project) -> None:
...
             file_node.ensure_uploaded(api=self)
 
-        response: Dict = requests.post(url=f"{self._host}/{project.node_type.lower()}", headers=self._http_headers, data=project.json).json()
+        node_known = len(self.search(type(node), SearchModes.UUID, str(node.uuid)).current_page_results) == 1
```
Comment (nh916) [inline src/cript/api/api.py:525]: I wrote a `get_bad_uuid_from_error_message` that I think would be helpful for getting the UUID from the error message

File: src/cript/api/api.py [hunk 5]
```diff
@@ -511,22 +510,42 @@ def save(self, project: Project) -> None:
...
 
+        json_data = node.get_json(known_uuid=known_uuid).json
+        if node_known:
+            response: Dict = requests.patch(url=f"{self._host}/{node.node_type.lower()}", headers=self._http_headers, data=json_data).json()
```
Comment (nh916) [inline src/cript/api/api.py:529]: I am guessing we'll have to send it in a specific order because otherwise we can send the project as patch and it would give us issues

File: src/cript/api/api.py [hunk 6]
```diff
@@ -511,22 +510,42 @@ def save(self, project: Project) -> None:
...
+        else:
+            response: Dict = requests.post(url=f"{self._host}/{node.node_type.lower()}", headers=self._http_headers, data=json_data).json()
+
+        if response["code"] == 400 and response["error"].startswith("Bad uuid:"):
```
Comment (nh916) [inline src/cript/api/api.py:533]: I catch this within the `brute_force_save` function and handle it there. I'd suggest using that if possible

File: src/cript/api/api.py [hunk 7]
```diff
@@ -511,22 +510,42 @@ def save(self, project: Project) -> None:
...
+            # We can have multiple occurrences of the node,
+            # but it doesn't matter which one we save
+            # TODO some error handling, for the BUG case of not finding the UUID
+            missing_node = node.find_children({"uuid": missing_uuid})[0]
+            known_uuid = self.save(missing_node, known_uuid)
+            known_uuid.add(missing_uuid)
+            # Recursive call to finish the post
+            return self.save(node, known_uuid)
+
+        print("asdf", response)
```
Comment (nh916) [inline src/cript/api/api.py:544]: does this need to be removed?

File: src/cript/nodes/core.py [hunk 8]
```diff
@@ -260,6 +261,9 @@ class ReturnTuple:
         previous_handled_nodes = copy.deepcopy(NodeEncoder.handled_ids)
         if handled_ids is not None:
             NodeEncoder.handled_ids = handled_ids
+        previous_known_uuid = copy.deepcopy(NodeEncoder.known_uuid)
```
Comment (nh916) [inline src/cript/nodes/core.py:264]: I advise against deepcopy whenever possible just because I had a very hard time when working with fixtures that used it for the integration tests. It just made things more complicated, but if we have to use it then we can use it

File: src/cript/nodes/util/__init__.py [hunk 9]
```diff
@@ -37,6 +38,14 @@ def default(self, obj):
                 if uid in NodeEncoder.handled_ids:
                     return {"node": obj._json_attrs.node, "uid": uid}
 
+            try:
+                uuid_str = str(obj.uuid)
```
Comment (nh916) [inline src/cript/nodes/util/__init__.py:42]: obj is a bad variable name, can we name it something more descriptive?

File: src/cript/nodes/util/__init__.py [hunk 10]
```diff
@@ -37,6 +38,14 @@ def default(self, obj):
...
+            except AttributeError:
```
Comment (nh916) [inline src/cript/nodes/util/__init__.py:43]: why are we catching an error and not doing anything with it?

**Comment by nh916** [comment]
honestly, I have a bit of a hard time understanding the full code

File: src/cript/api/paginator.py (hunk 1 above)
Comment (InnocentBug) [reply src/cript/api/paginator.py:203]: Well, the paginator was raising a cryptic exception when searching for a non-existent node.

But for the patch solution, we need it to return an empty list in that case.

I put it in a different commit, but I need this here to work.

File: src/cript/api/api.py (hunk 2 above)
Comment (InnocentBug) [reply src/cript/api/api.py:519]: This is just a name here.

It doesn't change anything for the user.

It is just for devs. For a dev it would be confusing to call it project, since it can be other things, when called differently.

I am ok, wrapping that a little differently, such that we expose one with `project` to the user, but we can do it after we confirm this works.

File: src/cript/api/api.py (hunk 3 above)
Comment (InnocentBug) [reply src/cript/api/api.py:522]: see comment above.

File: src/cript/api/api.py (hunk 4 above)
Comment (InnocentBug) [reply src/cript/api/api.py:525]: yeah, I didn't see or use that function. we can use it here.

Comment (InnocentBug) [reply src/cript/api/api.py:525]: well, actually no.

This is doing something else then you think.

This just checks if a UUID is known in the backend.
We can use a helper function for that too, if you like.

File: src/cript/api/api.py (hunk 5 above)
Comment (InnocentBug) [reply src/cript/api/api.py:529]: No, that shouldn't be necessary.
The only thing that I haven't handled yet, are `duplicate node` errors.
Everything else should automatically put itself in the right order.

File: src/cript/api/paginator.py (hunk 1 above)
Comment (nh916) [reply src/cript/api/paginator.py:203]: got it!

We have to come back to paginator later and really beef it up and refactor it because when I first wrote it it was just a rough draft of what I knew then

File: src/cript/api/api.py (hunk 6 above)
Comment (InnocentBug) [reply src/cript/api/api.py:533]: we can, use a helper function to catch this.
But your `brute_force_save` function, works fundamentally different then my save.

File: src/cript/api/api.py (hunk 7 above)
Comment (InnocentBug) [reply src/cript/api/api.py:544]: yeah, forget to remove.

File: src/cript/nodes/core.py (hunk 8 above)
Comment (InnocentBug) [reply src/cript/nodes/core.py:264]: No, there is a deepcopy necessary here.

However, this is a deepcopy of a list of strings. Not of nodes.
So very different then your issues in the fixtures.

The fixture issues are only caused, because DB schema doesn't allow UID where it should sometimes. We need to fix that separately.

File: src/cript/api/api.py (hunk 2 above)
Comment (nh916) [reply src/cript/api/api.py:519]: I agree, if we give the user to save only with project and we have a hidden one that works with node, that would be better. Yeah I agree we can do it after this works

File: src/cript/nodes/util/__init__.py (hunk 9 above)
Comment (InnocentBug) [reply src/cript/nodes/util/__init__.py:42]: This is the variable name that the documentation of the `json` module uses.
We should not deviate from that.

File: src/cript/nodes/util/__init__.py (hunk 10 above)
Comment (InnocentBug) [reply src/cript/nodes/util/__init__.py:43]: yes, that is a very pythonic way.

Instead of testing if an attribute is present, this is preferred.

File: src/cript/api/api.py (hunk 4 above)
Comment (nh916) [reply src/cript/api/api.py:525]: I do like helper functions!

Comment (nh916) [reply src/cript/api/api.py:525]: but I agree with you, lets get it working and then lets refactor it to make it better

File: src/cript/api/api.py (hunk 5 above)
Comment (nh916) [reply src/cript/api/api.py:529]: okie dokes, that sounds good. It feels too simple though

File: src/cript/api/api.py (hunk 6 above)
Comment (nh916) [reply src/cript/api/api.py:533]: okie dokes we can get something working for now and come back and refactor it after

File: src/cript/nodes/core.py (hunk 8 above)
Comment (nh916) [reply src/cript/nodes/core.py:264]: okie dokes, that sounds good!

File: src/cript/nodes/util/__init__.py (hunk 10 above)
Comment (nh916) [reply src/cript/nodes/util/__init__.py:43]: I agree, the principle of `try instead of checking for if`

File: src/cript/nodes/util/__init__.py (hunk 9 above)
Comment (nh916) [reply src/cript/nodes/util/__init__.py:44]: okie dokes, that makes sense

**Comment by nh916** [comment]
@InnocentBug made a small commit to fix the merge conflict issue

**Comment by nh916** [comment]
@InnocentBug I see a few errors on our side but mostly on API side, do we want to fix them before merging it in, or do you think this is good enough to approve for now and we can fix the rest later?

[Python-SDK-Test-Results-html-2023-07-06-17_06_26.pdf](https://github.com/C-Accel-CRIPT/Python-SDK/files/11973356/Python-SDK-Test-Results-html-2023-07-06-17_06_26.pdf)

**Comment by InnocentBug** [comment]
Yeah, there are a couple of errors left. 
I am working on fixing some of the errors but we can merge it before all errors are fixed.

**Review by nh916 — APPROVED** [review]
